base/inference.jl

```diff
@@ -2268,28 +2268,36 @@ function precise_container_type(@nospecialize(arg), @nospecialize(typ), vtypes::
     end
 end
 
+strip_iter_union(X::Type{Union{Nothing, T}}) where {T} = T
+strip_iter_union(::Type{Nothing}) = Nothing
+strip_iter_union(::Type{Union{}}) = Union{}
+strip_iter_union(T) = T
```

Comment:
this'll be very slow (because of compilation), and possibly unreliable (because `T = Union{Nothing, T}` is also a valid solution), use `typesubtract` instead